layerfetcher/layers.go

```diff
@@ -388,35 +387,40 @@ func (l *Logic) receiveLayerContent(ctx context.Context, layerID types.LayerID,
 // all blocks are fetched for a given layer.
 // it deliberately doesn't hold any lock while notifying channels.
 func notifyLayerBlocksResult(layerID types.LayerID, lyrDB layerDB, channels []chan LayerPromiseResult, lyrResult *layerResult, logger log.Log) {
-	var result *LayerPromiseResult
-	hasZeroBlockHash := false
-	var firstErr error
-	for _, res := range lyrResult.responses {
-		if res.err == nil && res.data != nil {
-			// at least one layer hash contains blocks. not a zero block layer
-			result = &LayerPromiseResult{Layer: layerID, Err: nil}
-			break
-		}
-		if res.err == ErrZeroLayer {
-			hasZeroBlockHash = true
-		} else if firstErr == nil {
-			firstErr = res.err
+	var (
+		result  LayerPromiseResult
+		missing int
+		err     error
+		highest types.LayerID
+	)
+	for bid, got := range lyrResult.blocks {
+		if !got {
+			logger.With().Warning("missing requested block", bid)
+			missing++
 		}
 	}
 
-	if result == nil { // no block data available
-		result = &LayerPromiseResult{Layer: layerID, Err: nil}
-		if hasZeroBlockHash {
-			// all other non-empty layer hashes returned errors. use the best information we've got
-			result.Err = ErrZeroLayer
-		} else {
-			// no usable result. just return the first error we received
-			result.Err = firstErr
+	if missing > 0 {
```

Comment:
can you try to simplify the code logic here a little bit? there is a fair amount of switch back and forth on the success and failure cases.

personally i don't see the value of report how many blocks missing, as long as one of the peer reported ErrBlockNotFetched

Reply:
i was using it during debugging to see if this is where blocks are missing, but right i can simply check all errors instead of one